File: 57. Insert Interval.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<Interval> insert(vector<Interval>& intervals, Interval newInterval) {
        if (intervals.empty())
            return vector<Interval>{newInterval};
        vector<Interval> res;
        int i;
        for(i = 0; i < intervals.size(); i++){
            if (intervals[i].end >= newInterval.start)
                break;
            res.push_back(intervals[i]);
        }
        if (i == intervals.size()){
            res.push_back(newInterval);
            return res;
        }
            
        if (intervals[i].start < newInterval.start)
            swap(newInterval, intervals[i]);
        res.push_back(newInterval);
        for(; i < intervals.size(); i++){
            if (res.back().end >= intervals[i].start)
                res.back().end = max(res.back().end, intervals[i].end);            
            else
                break;
        }
        for(; i < intervals.size();i++){
            res.push_back(intervals[i]);
        }
        return res;
    }
};

File: Python/57. Insert Interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        ans = list()
        added = False
        for interval in intervals:            
            if not added and interval[1] >= newInterval[0]:
                added = True
                if interval[0] <= newInterval[1]:
                    ans.append([min(interval[0], newInterval[0]), max(interval[1], newInterval[1])])
                if interval[0] > newInterval[1]:
                    ans.append(newInterval)
                    ans.append(interval)
            else:
                if len(ans) == 0 or interval[0] > ans[-1][1]:
                    ans.append(interval)
                else:
                    ans[-1][1] = max(interval[1], ans[-1][1])
        if not added:
            ans.append(newInterval)
        return ans
                    
                    
            